### PR TITLE
Create API for Alerts

### PR DESCRIPTION
**Pull request by dmonzonis**
Alerts have been refactored into their own model instead of being stored in a JSON field, and an API endpoint to retrieve alerts for all models has been created.

Closes OSIDB-325.

### CODE REVIEW

File: osidb/api_views.py
Comment (osoukup) [inline osidb/api_views.py:836]: `include_meta_attr_extend_schema_view` should probably go away

File: osidb/filters.py
Comment (osoukup) [inline osidb/filters.py:787]: I am thinking whether the `owner_` naming is good. At least for me it sounds like something about a person which is responsible for (resolving) the alert. I am thinking how to denote the relation by other name. What would you say about eg. `parent_` ?

Comment (JakubFrejlach) [inline osidb/filters.py:787]: Suggestion: This might be turned into a ChoiceFilter, I think we can retrieve all models which are subclassing some mixin like this approx:
```python
In [32]: def get_mixin_subclases(mixin):
    ...:     result = []
    ...:     for subclass in mixin.__subclasses__():
    ...:         if subclass._meta.abstract:
    ...:             result.extend(get_mixin_subclases(subclass))
    ...:         else:
    ...:             result.append(subclass.__name__)
    ...:     return result
    ...:

In [33]: get_mixin_subclases(AlertMixin)
Out[33]:
['Flaw',
 'Affect',
 'FlawCVSS',
 'AffectCVSS',
 'FlawComment',
 'FlawAcknowledgment',
 'FlawReference',
 'Package',
 'Snippet',
 'Tracker',
 'FlawMeta']
```

File: osidb/api_views.py
Comment (JakubFrejlach) [inline osidb/api_views.py:859]: If my suggestion with the owner model choices will be added I think this might be removed and drf-spectacular will pick up the choices from filter itself.

File: osidb/migrations/0123_alert_refactor.py
Comment (JakubFrejlach) [inline osidb/migrations/0123_alert_refactor.py:51]: Cannot apply the migration, I am getting prompted with a traceback on this particular line:
```
...
TypeError: Alert() got an unexpected keyword argument 'content_object'
```

**Review by JakubFrejlach — CHANGES_REQUESTED** [review]
Added some suggestions which are not really urgent, however locally I was not able to run the migration so not approving yet as not sure if migration is working or not.

File: openapi.yml
Comment (osoukup) [reply openapi.yml:7519]: It is not very intuitive and the docs creation is not even very documented as far as I can remember.

Comment (osoukup) [reply openapi.yml:7519]: The meaningful and understandable description for the user is the preference here. The developer can live with it :smile:

File: osidb/filters.py
Comment (dmonzonis) [reply osidb/filters.py]: To avoid computational overhead (if the subclasses have to be recomputed each time we get the choices) I have just added the possible options in the OpenApiParameter definition; this way we also get a list of options in the docs, although they will have to be revised in the event that the architecture changes.

File: osidb/mixins.py
Comment (dmonzonis) [reply osidb/mixins.py]: It was added mostly to not allow incorrect alert types, as there was a test case for that (`TestModelDefinitions::test_alert_incorrect_type`). But it's true that this case would be a programmer error since alerts are not created by users. So I will remove this test case as well.

File: osidb/filters.py
Comment (JakubFrejlach) [reply osidb/filters.py]: Sure, that is a good idea. That little function I included might go into `osidb/helpers.py` and it can be used to detect any changes from dev point of view. We already have some similar helper functions there as well (eg. get_unique_meta_attr_keys).

Comment (dmonzonis) [reply osidb/filters.py]: I will add it there. I did use a pretty similar function to get the list of models for the migration, but it was done ad-hoc,  didn't think about keeping it somewhere for future reference.

**Review by osoukup — APPROVED** [review]
LGTM

File: osidb/mixins.py
Comment (jsvob) [inline osidb/mixins.py:152]: Nitpick: I'd prefer to store the model itself in `osidb/models.py` for consistency with most other models.

Comment (dmonzonis) [reply osidb/mixins.py:152]: The only reason I added it here is to avoid a circular dependency: the `AlertMixin` uses `Alert`, and models in the `osidb/models.py` file use `AlertMixin`

Comment (osoukup) [inline osidb/mixins.py:152]: @dmonzonis I was thinking about using the alerts in the collector framework to store some data collection issues so we can query them easily. While thinking about it I realized one issue here.

The alerts do not have any ACL controls while they can still be freely queried on API by anyone authenticated. That means that even though a user would have only a public access-level, (s)he could eg. see alerts of the embargoed flaws and OSIDB would be leaking an information about their existence and potentially much more. The embargo is always taken quite seriously in ProdSec so I would see it as quite an issue.

I see two way how to overcome it. Probably easier way would be to restrict the API endpoint to those with the embargo access only. Anyone could see instance-specific alerts among its data so this general endpoint would have more narrowed usage anyway. A cleaner solution would be to provide the `Alert` model with ACLs which would be inherited from the parent instance.

I know this is quite late in game so feel free to create a separate Jira for this improvement and plan it for the next sprint so it does not interfere with your already planned work.

**Review by osoukup — CHANGES_REQUESTED** [review]

File: osidb/mixins.py
Comment (dmonzonis) [reply osidb/mixins.py]: I changed it so that alerts now properly use the parent's ACLs.

**Comment by dmonzonis** [comment]
@JakubFrejlach may I get a re-review from you as well? The original issues were fixed a while ago, and the only new change is that now `Alert` uses ACLs.

**Review by osoukup — APPROVED** [review]
LGTM. Thank you for addressing the ACL issue!

File: osidb/filters.py
Comment (JakubFrejlach) [inline osidb/filters.py:803]: I think this could be turned into `ChoiceFilter` and thus the whole query parameter annotation in the `osidb/api_views.py` might not be needed then but I have no strong preference and will leave it up to you.

**Review by JakubFrejlach — APPROVED** [review]
LGTM now! Great job.

File: osidb/migrations/0123_alert_refactor.py
Comment (JakubFrejlach) [reply osidb/migrations/0123_alert_refactor.py]: @dmonzonis So this was actually some problem on my side, I issued `make compose-down` and build everything from scratch and it works.

**Pull request merged**
